src/helpers/wsClient.ts

```diff
@@ -0,0 +1,178 @@
+
+export type WsType = 'ws' | 'wss';
+
+export type OnNewTransactionCallback = (
+  transaction: ChatMessageTransaction | TokenTransferTransaction
+) => void;
+
+export interface WsOptions {
+  /**
+   * ADM address to subscribe to notifications
+   */
+  admAddress: `U${string}`;
+
+  /**
+   * Websocket type: `'wss'` or `'ws'`. `'wss'` is recommended.
+   */
+  wsType?: WsType;
+
+  /**
+   * Must connect to node with minimum ping. Not recommended. Default is `false`.
+   */
+  useFastest?: boolean;
+
+  onNewMessage: OnNewTransactionCallback;
+}
+
+export class WebSocketClient {
+  /**
+   * Web socket client options.
+   */
+  public options: WsOptions;
+
+  /**
+   * Current socket connection
+   */
+  private connection?: Socket;
+
+  /**
+   * List of nodes that are active, synced and support socket.
+   */
+  private nodes: ActiveNode[];
+
+  private logger: Logger;
+  private onNewMessageCallback: OnNewTransactionCallback;
+
+  constructor(logger: Logger, options: WsOptions) {
+    this.logger = logger;
+    this.options = {
+      wsType: 'ws',
+      ...options,
+    };
+
+    this.nodes = [];
+    this.onNewMessageCallback = options.onNewMessage;
+  }
+
+  /**
+   * Filters nodes that support websocket and connects to one of them.
+   *
+   * @param nodes Sorted by ping array of active nodes
+   */
+  reviseConnection(nodes: ActiveNode[]) {
+    if (this.connection?.connected) {
+      return;
+    }
+
+    const {wsType} = this.options;
+
+    this.nodes = nodes.filter(
+      node =>
+        node.socketSupport &&
+        !node.outOfSync &&
+        // Remove nodes without IP if 'ws' connection type
+        (wsType !== 'ws' || !node.isHttps || node.ip)
+    );
+
+    this.setConnection();
+  }
+
+  /**
+   * Chooses node and sets up connection.
+   */
+  setConnection() {
+    const {logger} = this;
+
+    const supportedCount = this.nodes.length;
+    if (!supportedCount) {
+      logger.warn('[Socket] No supported socket nodes at the moment.');
+      return;
+    }
+
+    const node = this.chooseNode();
+    logger.log(
+      `[Socket] Supported nodes: ${supportedCount}. Connecting to ${node}...`
+    );
+    const connection = io(node, {
+      reconnection: false,
+      timeout: 5000,
+    });
+
+    connection.on('connect', () => {
+      const {admAddress} = this.options;
+
+      connection.emit('address', admAddress);
+      logger.info(
+        `[Socket] Connected to ${node} and subscribed to incoming transactions for ${admAddress}`
+      );
+    });
+
+    connection.on('disconnect', reason =>
+      logger.warn(`[Socket] Disconnected. Reason: ${reason}`)
+    );
+
+    connection.on('connect_error', error =>
+      logger.warn(`[Socket] Connection error: ${error}`)
+    );
+
+    connection.on('newTrans', transaction => {
+      if (transaction.recipientId !== this.options.admAddress) {
+        return;
+      }
+
+      if (
+        ![TransactionType.CHAT_MESSAGE, TransactionType.SEND].includes(
```

Comment:
Something like that.
@bludnic Suggestions?

Reply:
Proposed solution looks good.

We also can add helpers like `.onMessage()`, `.onTransfer()` that will just wrap `.onTransaction([...])`.

And I think it will be great to have a possibility to unsubscribe from that events.

Reply:
I mean this:

```js
const socket = new WebSocketClient({ admAddress: 'U1234...' })

function onMessage(transaction) {
  console.log(transaction)
}

socket.on(TransactionType.MESSAGE, onMessage)
socket.off(onMessage) // unsubscribe
```

I used non-anonymous `onMessage` function to keep the same reference when unsubscribing.

Reply:
The `socket.transfer` may be confusing, like it's some kind of action, and not an event. I think its better to prefix with `on`.

Reply:
Example subscribing multiple times:

```js
const socket = new WebSocketClient({ admAddress: 'U1234...' })

function onMessage(transaction) {
  console.log(transaction)
}

function onMessage2(transaction) {
  console.log(transaction)
}

socket.on(TransactionType.MESSAGE, onMessage)
socket.on(TransactionType.MESSAGE, onMessage2)
socket.off(onMessage)
socket.off(onMessage2)
```